hadoop-hdfs-project/hadoop-hdfs/src/test/java/org/apache/hadoop/hdfs/TestByteBufferPread.java

```diff
@@ -161,130 +229,264 @@ private void testPreadWithFullByteBuffer(ByteBuffer buffer)
-   * contents are properly loaded into the supplied {@link ByteBuffer}.
+   * contents are properly loaded into the {@link ByteBuffer}.
    */
-  private void testPreadFullyWithByteBuffer(ByteBuffer buffer)
-          throws IOException {
+  @Test
+  public void testPreadFullyWithByteBuffer() throws IOException {
     int totalBytesRead = 0;
     try (FSDataInputStream in = fs.open(testFile)) {
       in.readFully(totalBytesRead, buffer);
       // Make sure the buffer is full
-      assertFalse(buffer.hasRemaining());
+      assertBufferIsFull();
       // Make sure the contents of the read buffer equal the contents of the
       // file
-      buffer.position(0);
-      byte[] bufferContents = new byte[FILE_SIZE];
-      buffer.get(bufferContents);
-      assertArrayEquals(bufferContents, fileContents);
+      assertBufferEqualsFileContents(0, FILE_SIZE, 0);
+    }
+  }
+
+  /**
+   * readFully past the end of the file into an empty buffer; expect this
+   * to be a no-op.
+   */
+  @Test
+  public void testPreadFullyPastEOFEmptyByteBuffer() throws IOException {
+    try (FSDataInputStream in = fs.open(testFile)) {
+      in.readFully(FILE_SIZE + 10, emptyBuffer);
+    }
+  }
+
+  /**
+   * Reads from a negative position -expects a failure.
+   * Also uses the new openFile() API to improve its coverage.
+   */
+  @Test
+  public void testPreadFullyNegativeOffset() throws Exception {
+    try (FSDataInputStream in = fs.openFile(testFile).build().get()) {
+      in.seek(QUARTER_SIZE);
+      intercept(EOFException.class, NEGATIVE_POSITION_READ,
+          () -> in.readFully(-1, buffer));
+      // the stream position has not changed.
+      assertStreamPosition(in, QUARTER_SIZE);
+    }
+  }
+
+  /**
+   * Read fully with a start position past the EOF -expects a failure.
+   */
+  @Test
+  public void testPreadFullyPositionPastEOF() throws Exception {
+    try (FSDataInputStream in = fs.openFile(testFile).build().get()) {
+      in.seek(QUARTER_SIZE);
+      intercept(EOFException.class, EOF_IN_READ_FULLY,
+          () -> in.readFully(FILE_SIZE * 2, buffer));
+      // the stream position has not changed.
+      assertStreamPosition(in, QUARTER_SIZE);
+    }
+  }
+
+  /**
+   * Read which goes past the EOF; expects a failure.
+   * The final state of the buffer is undefined; it may fail fast or fail late.
+   * Also uses the new openFile() API to improve its coverage.
+   */
+  @Test
+  public void testPreadFullySpansEOF() throws Exception {
+    try (FSDataInputStream in = fs.openFile(testFile).build().get()) {
+      intercept(EOFException.class, EOF_IN_READ_FULLY,
+          () -> in.readFully(FILE_SIZE - 10, buffer));
+      if (buffer.position() > 0) {
+        // this implementation does not do a range check before the read;
+        // it got partway through before failing.
+        // this is not an error -just inefficient.
+        LOG.warn("Buffer reads began before range checks with {}", in);
```

Comment:
What is the benefit of having this, is this log gonna help sometime is debugging?